### PR TITLE
Use AWS for Ephemeral Dev Environments

### PR DESCRIPTION
**Pull request by cindyledev**
Related to #3024 

This PR adds step-by-step instructions on how to set up a development environment to develop Telescope remotely on an AWS EC2 instance. Please review src/docs/docs/getting-started/vscode-ssh.md of this PR

### CODE REVIEW

**Review by humphd — COMMENTED** [review]
Screen Shot 2022-02-27 at 3 21 18 PM

I followed this doc step-by-step and was able to get things running.  Amazing!

I'll do a proper review this week, but overall, it looks good.  I think we could probably automate the IP address stuff for working on the remote instance vs. localhost.

I was interested to see how much space the install would take, and I'm using 14G of my 40G (`df -h`).

I also installed the latest GitHub CLI, which I think we should add to your init script.  I used `sudo yum install -y https://github.com/cli/cli/releases/download/v2.5.1/gh_2.5.1_linux_amd64.rpm`.  The most recent release is at https://github.com/cli/cli/releases.  Once you authenticate (`gh auth login`) as your GitHub account, you can use `gh` to do things like checkout a PR with `gh pr checkout 3055` and test it easily.

It's pretty cool how VSCode manages your port forwarding:

Screen Shot 2022-02-27 at 3 37 18 PM

And you can configure/see them all at once:

Screen Shot 2022-02-27 at 3 37 46 PM

I tried binding port 80 on ec2 to my local 53081 and I can see all the remote services:

Screen Shot 2022-02-27 at 3 41 07 PM

In other words, this is all going to work.  We can just smooth out the process a bit.

Amazing.

**Comment by humphd** [comment]
Another thing we should do is add the [VSCode Docker Extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker) to our `.vscode/extensions.json` and figure out how to auto-install our recommended extensions in the remote VSCode instance running on ec2.  Then you can graphically manage and interact with the containers

I seem to have to manually install things into the remote ec2 server, I'm not sure how to force this to happen by default.

**Comment by cindyledev** [comment]
> Another thing we should do is add the [VSCode Docker Extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker) to our `.vscode/extensions.json` and figure out how to auto-install our recommended extensions in the remote VSCode instance running on ec2. Then you can graphically manage and interact with the containers
> 
> I seem to have to manually install things into the remote ec2 server, I'm not sure how to force this to happen by default.

I looked into this and we can meet this requirement by using [dev containers](https://code.visualstudio.com/docs/remote/create-dev-container). Dev containers lets us use a Docker container as a full-featured development environment which deviates from the approach we have taken here. I think Dev containers will allow us to achieve more of what we want:
- forcing extensions to be installed inside the dev container
- force forward ports

We know what have in this PR is working and is `Good`. Dev containers could potentially be `Better`. It will take a bit of time to get the dev containers working so I think I can clean up this PR so we can review it. Once accepted, we will have something `Good` then in a follow-up PR, we can do `Better`.

**Comment by humphd** [comment]
I agree, let's leave dev containers for later, if ever.

**Comment by cindyledev** [comment]
This is ready for review. Tagging @tpmai22 @JerryHue @TueeNguyen @dbelokon and @humphd

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 1]
```diff
@@ -1,51 +1,61 @@
...
-- Step 4 - Add Storage: Change the Size (GiB) from `8` to `20`
-  ![](https://seneca-cdot-telescope.s3.amazonaws.com/vscode-ssh/2021-10-26+12_51_25-vscode-ssh.md+-+telescope+-+Visual+Studio+Code.png)
+- Step 1 - Choose an Amazon Machine Image (AMI): `Amazon Linux 2 AMI (HVM) - Kernel 5.10, SSD Volume Type`
+  ![AWS Linux 2 AMI](../../static/img/aws_linux_2_ami.png)
+- Step 2 - Choose an Instance Type: `r5.large`
+  ![AWS r5.large Instance Type](../../static/img/aws_instance_type.png)
+- Step 3 - Configure Instance Details:
+  - IAM role: LabInstanceProfile
+    ![AWS LabInstanceProfile IAM Role](../../static/img/aws_iam_role.png)
+  - Copy and paste the [aws-userdata.sh](../../../../tools/aws-userdata.sh) script into the `User data` field
```
Comment (humphd) [inline src/docs/docs/getting-started/vscode-ssh.md:55]: Question: why not include as a file?

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 2]
```diff
@@ -1,51 +1,61 @@
...
+    ![AWS Userdata](../../static/img/aws_userdata.png)
+- Step 4 - Add Storage: Change the Size (GiB) from `8` to `40`
```
Comment (humphd) [inline src/docs/docs/getting-started/vscode-ssh.md:57]: Question: why 40?  I think it's fine, but I'm curious.

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 3]
```diff
@@ -73,47 +83,67 @@ The following will show you how to create and connect to a virtual machine (VM)
     - Port Range: `8443`
     - Source: `My IP`
 
-  ![](https://seneca-cdot-telescope.s3.amazonaws.com/vscode-ssh/2021-10-26+12_54_12-vscode-ssh.md+-+telescope+-+Visual+Studio+Code.png)
+  ![AWS Security Groups](https://seneca-cdot-telescope.s3.amazonaws.com/vscode-ssh/2021-10-26+12_54_12-vscode-ssh.md+-+telescope+-+Visual+Studio+Code.png)
```
Comment (humphd) [inline src/docs/docs/getting-started/vscode-ssh.md:86]: Are we going to run into issues with people who don't have a static IP and how we're telling them to do their security group (i.e., using My IP if it changes between sessions)?

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 4]
```diff
@@ -309,7 +238,9 @@ sh ./tools/aws-ip.sh
 pnpm install
 ```
 
-2. Start all Telescope services. This will take some time to complete
+2. Start all Telescope services using the environment variables set in `.env`. This will take some time to complete
+
+Note: Do not use `pnpm services:start`. This will use the environment variables in `config/env.development` and we don't want that here.
```
Comment (humphd) [inline src/docs/docs/getting-started/vscode-ssh.md:243]: We should fix that in a follow-up.  We probably need to create some configs for running within AWS.

**Review by humphd — CHANGES_REQUESTED** [review]
Fantastic.  A few things.

File: tools/aws-userdata.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+sudo systemctl enable docker
+sudo usermod -a -G docker ec2-user
+
+# Install docker-compose
+wget https://github.com/docker/compose/releases/latest/download/docker-compose-$(uname -s)-$(uname -m)
+sudo mv docker-compose-$(uname -s)-$(uname -m) /usr/local/bin/docker-compose
+sudo chmod -v +x /usr/local/bin/docker-compose
+
+# Install Node.js
+curl -sL https://rpm.nodesource.com/setup_16.x | sudo bash -
```
[the remark this answers is not in the document]
Comment (cindyledev) [reply tools/aws-userdata.sh]: Updated to `lts`

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 6]
```diff
@@ -1,51 +1,61 @@
...
+- Recommended: `r5.large (16 GiB RAM + 2 vCPU)`
 
 **Summary of Pricing**:
 
-- t2.medium costs \$0.0464 per hour
 - t2.large costs \$0.0928 per hour
-- 30GB Amazon Elastic Block Storage (EBS) costs \$3 per month
+- r5.large costs \$0.126 per hour
+- 40GB Amazon Elastic Block Storage (EBS) costs \$1 per month
+- [Elastic IP Address](https://aws.amazon.com/ec2/pricing/on-demand/#Elastic_IP_Addresses)
```
[the remark this answers is not in the document]
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md:19]: Added cost of Elastic IP address

File: src/docs/docs/getting-started/vscode-ssh.md (hunk 1 above)
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md:56]: I thought it'd be easier to copy and paste but using a file would yield the same result. I figured it would take more steps to download the file from the repo, save it to a folder, then upload it... I use Windows... there's probably a better way to do this on Mac XD

File: src/docs/docs/getting-started/vscode-ssh.md (hunk 2 above)
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md:58]: When I was playing around with the storage size for Cloud9 (it initially gave me 10GB), I upgraded it to 20GB and ran out of space after building all of our containers so I just thought 40GB would guarantee me not to have problems. In hindsight, Cloud9 probably has other stuff pre-installed so that's why it takes up more storage... I know our containers take up almost 10GB and Telescope repo takes up over 2GB after `pnpm install`... HMMMM Maybe 25 GB is good enough here XD I will change it.

File: src/docs/docs/getting-started/vscode-ssh.md (hunk 3 above)
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md:87]: I put a note stating that if their IP address changes, they should update their Security Group. Common causes of IP address changes are:
- changing internet providers
- route rebooting (either manually or power outage)

I want to somehow teach everyone about security groups and IP addresses because they are often the culprits when something goes wrong...

File: src/docs/docs/getting-started/vscode-ssh.md [hunk 7]
```diff
@@ -167,132 +176,52 @@ aws configure
...
+4. Generate a new SSH key to add to your GitHub account? `Yes`
+5. Enter a passphase for your new SSH key (Optional): `********`
+6. How would you like to authenticate GitHub CLI? `Login with a web browser`
+7. First copy your one-time code: `ABC1-234D`
 
-## Install Node.js
+## Opening up the Telescope repository in AWS EC2:
 
-1. Install Node.js 16.x
+1. Clone your fork of the Telescope repository. For example `gh repo clone cindyledev/telescope`
```
[the remark this answers is not in the document]
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md]: I changed this to
```
1. Clone your fork of the Telescope repository. For example `gh repo clone cindyledev/telescope`. If you do not have a fork of the Telescope repository, run `gh repo clone -o upstream Seneca-CDOT/telescope` to clone the Telescope repository and name the remote `upstream` then proceed to Step 5.
```

File: src/docs/docs/getting-started/vscode-ssh.md (hunk 4 above)
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md:246]: This is related to #3017 , we can add on to it or file a new issue. I'd like to see how the team resolves #3017 so I will follow it closely.

File: src/docs/docs/getting-started/vscode-ssh.md (hunk 2 above)
Comment (cindyledev) [reply src/docs/docs/getting-started/vscode-ssh.md]: I double-checked with `df -h` and apparently, we only use 14GB for everything which I think you mentioned somewhere before. Forgive me, it totally slipped my mind. I changed it to 20GB

**Review by JerryHue — COMMENTED** [review]
Since I recently moved my environment, I am not using Visual Studio Code anymore for development, so I won't be able to test this thoroughly :(

However, I would like to praise the documentation, since it was very detailed and easy to follow. Nice job, @cindyledev!

**Review by humphd — DISMISSED** [review]
Simply amazing, thank you!

**Review by tpmai22 — DISMISSED** [review]
Great work !

**Review by tpmai22 — APPROVED** [review]
Somehow rerun the test dismiss all the reviews ! So here we go again !
